### ISSUE TITLE
shipping as cart item

### ISSUE THREAD

**Issue by benseitz** [kirby-kart/kart#13]
I just thought: shipping fees are nothing but a special article. They have a price, a quantity (if more than one shippment is needed), a tax and everything else any other cart item has.

A shipping extension could go through the item(s) and then decide which shipping item to add to the cart. There could be different shipping variations and everything.

What do you think?

**Comment by distantnative** [comment]
I totally agree. Even more, I think taxes could be handled similarly: each tax category is a special article, the tax rate a (admittedly weird) quantity... okay, taxes would have their special 0-taxes-on-taxes tax .

In the starterkit, I'm laying it out as follows:

```php
<ul>
    <?php foreach ($cart->items() as $item) : ?>
        <li>
            <?php echo $item->product()->title(); ?>
            <?php echo $item->quantity(); ?>
            <?php echo $item->surcharges(); ?>
            <?php echo $item->line_total(); ?>
        </li>
    <?php endforeach; ?>
    <?php foreach ($cart->surcharges() as $surcharge) : ?>
        <li>
            <?php echo $surcharge->title(); ?>
            <?php echo $surcharge->amount(); ?>
        </li>
    <?php endforeach; ?>
    <li>TOTAL <?php echo $cart->cart_total(); ?></li>
</ul>
```

The idea would be that extensions such as shipping and taxes hook into the ```->surcharges()``` method which then returns an array of special items (e.g. 2 tax lines + 1 shipping line).

So a cart in that concept would contain of normal product items (```->items()```) and special surcharge items (```->surcharges()```).

**Comment by plaidscreen** [comment]
Often shipping charge is based on location of the recipient, at least it is here in the US, where all the major parcel services have 8 different shipping "zones" that change the shipper's cost. However I still think it makes sense that the shipping extension would hook into the Cart, instead of the Order. It'll just have to grab info from both.

**Comment by plaidscreen** [comment]
@distantnative That looks great - Exactly how I was envisioning it.

**Comment by distantnative** [comment]
Ahh doh, I was always thinking about the shipping extensions in a way that they have to calculate the charge on location... and then forgot that the address kinda doesn't get provided before checkout. So there'd be the need to have at least a 2-step checkout, where first the address is entered so that in the second step the final amount wit correct shipping charge can be displayed.

**Comment by benseitz** [comment]
At least in germany theres the gross/net (brutto/netto) proplem
b2b shops show the article price as net, b2c shops as gross

**Comment by plaidscreen** [comment]
@distantnative I often see shopping carts that allow you to enter a postal code before checkout, to get a shipping "estimate" before the full address is entered. Of course, we could leave this feature up to the theme developer.

**Comment by benseitz** [comment]
leaving it up to the theme developer seems like a good solution if possible

**Comment by distantnative** [comment]
@benseitz but wouldn't be that set up in the configs?

@plaidscreen Yea, I think so too. Shipping extension should maybe throw a kind of error when called through the surcharges-hook and no address is available, which then could still appear in the surcharges-return-array and displayed.

**Comment by benseitz** [comment]
in  b2b shops (or all american shops) the article price is net - surcharges would add a percentage to that.
in b2c shops (at least german ones) the article price is gross - surcharges would show the tax part of the price

And I don't know about any other countries, but taxes seem to be crazy... Maybe it would be best to have a special tax-object and then have special tax-extensions for each countries?
Without proper tax support a online shop is unusable in certain countries.

**Comment by distantnative** [comment]
Or defining a structure of tax elements that can be relatively easily set up be the admin:

```php
$tay = array(
    'taxID'  => 'vat19',
    'title'  => 'VAT (19%)',
    'rate'   => 0.19,
    'target' => 'cart' // or 'item'
    'part'   => 'net'
)
```

Just as a basis idea, prob a few more factors would be needed. But something that could potentially cover all ways taxes can be applies. And the admin then just sets rules like this one and the tax extension can calculate on basis of this all the taxes for items/carts, gross/net...?

**Comment by distantnative** [comment]
Additionally, it would be great to provide a resource of common tax rules for countries. But also from a legal standpoint, I think it would make sense to let the admin, who sets the shop up, define the tax rates and where and when they are applied.

**Comment by benseitz** [comment]
I agree, letting the admin take care of that sounds good.
